util/cpuprofile/cpuprofile.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cpuprofile
+
+import (
+	"bytes"
+	"errors"
+	"runtime/pprof"
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/metrics"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/atomic"
+)
+
+var defProfileDuration = time.Second
+
+// globalCPUProfiler is the global CPU profiler.
+// If you want to create a new global cpu profiler, you should stop the old global cpu profiler.
+var globalCPUProfiler = newParallelCPUProfiler()
+
+// ProfileConsumer is the profile data consumer.
+// ProfileConsumer is a channel alias, if the channel is full, then the channel won't receive the latest profile data
+// until it is not blocked.
+type ProfileConsumer = chan *ProfileData
+
+// ProfileData contains the cpu profile data and some additional information.
+type ProfileData struct {
+	Data  *bytes.Buffer
+	Begin time.Time
+	End   time.Time
+	Error error
+}
+
+// StartCPUProfiler uses to start to run the global parallelCPUProfiler.
+func StartCPUProfiler() error {
+	return globalCPUProfiler.start()
+}
+
+// CloseCPUProfiler uses to close the global parallelCPUProfiler.
+func CloseCPUProfiler() {
+	globalCPUProfiler.close()
+}
+
+// Register register a ProfileConsumer into the global CPU profiler.
+// Normally, the registered ProfileConsumer will receive the cpu profile data per second.
+// If the ProfileConsumer (channel) is full, the latest cpu profile data will not be sent to it.
+// This function is thread-safe.
+func Register(ch ProfileConsumer) {
+	globalCPUProfiler.register(ch)
+}
+
+// Unregister unregister a ProfileConsumer from the global CPU profiler.
+// The unregistered ProfileConsumer won't receive the cpu profile data any more.
+// This function is thread-safe.
+func Unregister(ch ProfileConsumer) {
+	globalCPUProfiler.unregister(ch)
+}
+
+// ConsumersCount returns the count of the global parallelCPUProfiler's consumer.It is exporting for test.
+func ConsumersCount() int {
+	return globalCPUProfiler.consumersCount()
+}
+
+// parallelCPUProfiler is a cpu profiler.
+// With parallelCPUProfiler, it is possible to have multiple profile consumer at the same time.
+// WARN: Only one running parallelCPUProfiler is allowed in the process, otherwise some profiler may profiling fail.
+type parallelCPUProfiler struct {
+	sync.Mutex
+	cs     map[ProfileConsumer]struct{}
+	notify chan struct{}
```

Comment:
What does it notify? It would be great if this can be known from the name instead of reading the code thoroughly.